File: java/src/eu/semaine/components/control/ParticipantControl.java
/**
 * Copyright (C) 2008 DFKI GmbH. All rights reserved.
 * Use is subject to license terms -- see license.txt.
 */
package eu.semaine.components.control;

import java.util.HashMap;
import java.util.Map;

import javax.jms.JMSException;

import eu.semaine.components.Component;
import eu.semaine.datatypes.stateinfo.ContextStateInfo;
import eu.semaine.datatypes.stateinfo.StateInfo;
import eu.semaine.jms.message.SEMAINEMessage;
import eu.semaine.jms.message.SEMAINEStateMessage;
import eu.semaine.jms.receiver.StateReceiver;
import eu.semaine.jms.sender.StateSender;
import eu.semaine.system.CharacterConfigInfo;

/**
 * @author marc
 *
 */
public class ParticipantControl extends Component 
{

    private ParticipantControlGUI gui;
	private StateSender contextSender;
	private StateReceiver contextReceiver;
    private boolean isUserPresent = false;
    private String currentCharacter = CharacterConfigInfo.getDefaultCharacter().getName();
    private String nextCharacter = null;
    private String currentDialogContext = null;

    /**
	 * @param componentName
	 * @throws JMSException
	 */
	public ParticipantControl() throws JMSException 
	{
		super("ParticipantControlGUI", true, false);
		contextSender = new StateSender("semaine.data.state.context", StateInfo.Type.ContextState, getName());
		senders.add(contextSender); // so it can be started etc
        contextReceiver = new StateReceiver("semaine.data.state.context", StateInfo.Type.ContextState);        
		receivers.add(contextReceiver);
        gui = new ParticipantControlGUI(this);
        gui.setVisible(true);
	}
	
	@Override
	protected void act() throws JMSException {
		if (hasSystemJustBecomeReady()) {
			sendWhoIsPresent();
		}
	}

        @Override
        protected void react(SEMAINEMessage m)
        throws Exception
        {
            SEMAINEStateMessage xm = (SEMAINEStateMessage) m;
            StateInfo state = xm.getState();
            if (state.hasInfo("userPresent")) {
            	isUserPresent = "present".equals(state.getInfo("userPresent"));
            }
            if (state.hasInfo("character")) {
            	currentCharacter = state.getInfo("character");
            }
            if (state.hasInfo("nextCharacter")) {
            	nextCharacter = state.getInfo("nextCharacter");
            }
            if (state.hasInfo("currentDialogContext")) {
            	currentDialogContext = state.getInfo("currentDialogContext");
            }
            gui.updateWhoIsPresent();
        }

	private void sendWhoIsPresent()
	throws JMSException
	{
		Map<String, String> info = new HashMap<String, String>();
		info.put("userPresent", isUserPresent ? "present" : "absent");
		info.put("character", currentCharacter);
		if (nextCharacter != null) {
			info.put("nextCharacter", nextCharacter);
		}
		if (currentDialogContext != null) {
			info.put("dialogContext", currentDialogContext);
		}
		ContextStateInfo context = new ContextStateInfo(info);
		contextSender.sendStateInfo(context, meta.getTime());
	}

    public boolean isUserPresent()
    {
        return isUserPresent;
    }
    
    public void setUserPresent(boolean userPresent)
    {
        this.isUserPresent = userPresent;
        try {
            sendWhoIsPresent();
        } catch (JMSException e) {
            e.printStackTrace();
        }
    }
    
    public String getCurrentCharacter()
    {
        return currentCharacter;
    }

    public String getNextCharacter() {
    	return nextCharacter;
    }
    
    public String getDialogContext() {
    	return currentDialogContext;
    }
    
    public void setCurrentCharacter(String name)
    {
    	setCharacterInfo(name, null, null);
    }
    
    public void setCharacterInfo(String name, String nextChar, String dialogContext) {
    	this.currentCharacter = name;
    	this.nextCharacter = nextChar;
    	this.currentDialogContext = dialogContext;
        try {
            sendWhoIsPresent();
        } catch (JMSException e) {
            e.printStackTrace();
        }
    }
    
}

File: c++/src/semaine/components/control/ParticipantControl.cpp
/*
 * ParticipantControl.cpp
 *
 *  Created on: 28.07.2009
 *      Author: marc
 */

#include <semaine/components/control/ParticipantControl.h>

#include <semaine/cms/message/SEMAINEStateMessage.h>
#include <semaine/datatypes/stateinfo/ContextStateInfo.h>

#include <decaf/lang/Thread.h>

#include <string>
#include <iostream>

using namespace cms;
using namespace semaine::cms::sender;
using namespace semaine::cms::receiver;
using namespace semaine::cms::message;
using namespace semaine::datatypes::stateinfo;

namespace semaine {

namespace components {

namespace control {


class StdinLineReader : public decaf::lang::Thread
{
public:
	StdinLineReader(ParticipantControl * pc)
	: pc(pc), exitRequested(false) {}

	void run() {
		std::string line;
		while (!exitRequested) {
			getline(std::cin, line);
			if (std::cin.eof()) {
				exitRequested = true;
				pc->requestExit();
			} else {
				std::string user = "";
				std::string character = "";
				if (line.find("user") != std::string::npos) {
					if (line.find("present") != std::string::npos) user = "present";
					else user = "absent";
				}
				if (line.find("Poppy") != std::string::npos) character = "Poppy";
				else if (line.find("Prudence") != std::string::npos) character = "Prudence";
				else if (line.find("Obadiah") != std::string::npos) character = "Obadiah";
				else if (line.find("Spike") != std::string::npos) character = "Spike";
				pc->sendContextInfo(user, character);
			}
		}
	}
private:
	ParticipantControl * pc;
	bool exitRequested;
};


ParticipantControl::ParticipantControl()
throw (CMSException)
: semaine::components::Component("ParticipantControlC++", false, false)
{
	contextSender = new StateSender("semaine.data.state.context", StateInfo::ContextState, getName());
	senders.push_back(contextSender);
	contextReceiver = new StateReceiver("semaine.data.state.context", StateInfo::ContextState);
	receivers.push_back(contextReceiver);
}

ParticipantControl::~ParticipantControl() {
}

void ParticipantControl::react(SEMAINEMessage * message) throw (std::exception)
{
	SEMAINEStateMessage * sm = dynamic_cast<SEMAINEStateMessage *>(message);
	if (sm == NULL) throw MessageFormatException(std::string("Expected state message, got a ")+typeid(*message).name());
	StateInfo * state = sm->getState();
	// Either like this:
	//std::string userPresent = state->getInfo("userPresent");
	//std::string character = state->getInfo("character");
	std::cout << "Received context message: ";
	// Or, more flexibly, like this:
	std::string values[] = {"userPresent", "character"};
	int n = 2;
	for (int i=0; i<n; i++) {
		if (state->hasInfo(values[i])) std::cout << " " << values[i] << "=" << state->getInfo(values[i]);
	}
	std::cout << std::endl;
}

void ParticipantControl::customStartIO() throw(std::exception)
{
	StdinLineReader * slr = new StdinLineReader(this);
	slr->start();
}

void ParticipantControl::sendContextInfo(const std::string & userPresent, const std::string & character)
{
	std::map<std::string, std::string> info;
	bool haveSomething = false;
	if (userPresent == "present" || userPresent == "absent") {
		info["userPresent"] = userPresent;
		haveSomething = true;
	}
	if (character == "Spike" || character == "Poppy" || character == "Prudence" || character == "Obadiah") {
		info["character"] = character;
		haveSomething = true;
	}
	if (haveSomething) {
		ContextStateInfo csi(info);
		contextSender->sendStateInfo(&csi, meta.getTime());
	}
}




}

}

}
